akka-stream-tests/src/test/scala/akka/stream/remote/StreamRefsSpec.scala

```diff
@@ -0,0 +1,290 @@
+      //          SinkRef.bulkTransferSource()
+      //            .to(Sink.actorRef(probe, "<COMPLETE>"))
+      //            .run()
+      //
+      //
+      //        sender() ! BulkSinkMsg(sink)
+    }
+
+  }
+
+  // -------------------------
+
+  final case class SourceMsg(dataSource: SourceRef[String])
+  final case class BulkSourceMsg(dataSource: SourceRef[ByteString])
+  final case class SinkMsg(dataSink: SinkRef[String])
+  final case class BulkSinkMsg(dataSink: SinkRef[ByteString])
+
+  def config(): Config = {
+    val address = SocketUtil.temporaryServerAddress()
+    ConfigFactory.parseString(
+      s"""
+    akka {
+      loglevel = INFO
+
+      actor {
+        provider = remote
+        serialize-messages = off
+      }
+
+      remote.netty.tcp {
+        port = ${address.getPort}
+        hostname = "${address.getHostName}"
+      }
+    }
+  """).withFallback(ConfigFactory.load())
+  }
+}
+
+class StreamRefsSpec(config: Config) extends AkkaSpec(config) with ImplicitSender {
+  import StreamRefsSpec._
+
+  def this() {
+    this(StreamRefsSpec.config())
+  }
+
+  val remoteSystem = ActorSystem("RemoteSystem", StreamRefsSpec.config())
+  implicit val mat = ActorMaterializer()
+
+  override protected def beforeTermination(): Unit =
+    TestKit.shutdownActorSystem(remoteSystem)
+
+  val p = TestProbe()
+
+  // obtain the remoteActor ref via selection in order to use _real_ remoting in this test
+  val remoteActor = {
+    val it = remoteSystem.actorOf(DatasourceActor.props(p.ref), "remoteActor")
+    val remoteAddress = remoteSystem.asInstanceOf[ActorSystemImpl].provider.getDefaultAddress
+    system.actorSelection(it.path.toStringWithAddress(remoteAddress)) ! Identify("hi")
+    expectMsgType[ActorIdentity].ref.get
+  }
+
+  "A SourceRef" must {
+
+    "send messages via remoting" in {
+      remoteActor ! "give"
+      val sourceRef = expectMsgType[SourceRef[String]]
+
+      sourceRef
+        .runWith(Sink.actorRef(p.ref, "<COMPLETE>"))
+
+      p.expectMsg("hello")
+      p.expectMsg("world")
+      p.expectMsg("<COMPLETE>")
+    }
+
+    "fail when remote source failed" in {
+      remoteActor ! "give-fail"
+      val sourceRef = expectMsgType[SourceRef[String]]
+
+      sourceRef
+        .runWith(Sink.actorRef(p.ref, "<COMPLETE>"))
+
+      val f = p.expectMsgType[Failure]
+      f.cause.getMessage should include("Remote stream (")
+      // actor name here, for easier identification
+      f.cause.getMessage should include("failed, reason: Booooom!")
+    }
+
+    "complete properly when remote source is empty" in {
+      // this is a special case since it makes sure that the remote stage is still there when we connect to it
+
+      remoteActor ! "give-complete-asap"
+      val sourceRef = expectMsgType[SourceRef[String]]
+
+      sourceRef
+        .runWith(Sink.actorRef(p.ref, "<COMPLETE>"))
+
+      p.expectMsg("<COMPLETE>")
+    }
+
+  }
+
+  "A SinkRef" must {
+
+    "xoxo receive elements via remoting" in {
```

Comment:
xoxo ?

Reply:
That's my pattern for using in test only: `testOnly *StreamRefs* -- -z xoxo`
Will remove before merge ofc